### PR TITLE
[PT] New translation: Mox

### LINKED ISSUE #2391: [PT] New translation: Mox
## Motivation

The portuguese translation for the lesson [Mox](https://elixirschool.com/en/lessons/libraries/mox/) is missing.

## Proposed Solution

- First create a new file under the `pt` folder with the content of the lesson.

`cp en/lessons/libraries/mox.md pt/lessons/libraries/mox.md`

- Then translate the file at `pt/lessons/libraries/mox.md`

## Tips

- Check the [contributing guidelines](https://github.com/elixirschool/elixirschool/blob/master/CONTRIBUTING.md) before starting.
- Feel free to open a [draft PR](https://github.blog/2019-02-14-introducing-draft-pull-requests/#tag-your-work-in-progress) as soon as you start working on the translation. In order to prevent we prevent having multiple people working on the translation and also allow the reviewers to give early feedback.
- In case of any doubt feel free to comment in this issue.

### PR DESCRIPTION
**Pull request by ItzaMi**
Closes #2391 

I'm Portuguese from Portugal and I did my best to translate it according to the rest of the website which seems, mostly, Portuguese from Brazil.

Some words I didn't translate because it didn't make sense to me to translate stuff like `third-party` or even the word `mocks`.

Let me know if something needs to be changed 😄 And thank you for letting me be part of this 💪

### CODE REVIEW

File: pt/lessons/libraries/mox.md [hunk 1]
```diff
@@ -0,0 +1,222 @@
+---
+version: 1.0.0
+title: Mox
+---
+
+Mox é uma biblioteca feita para projectar mocks simultâneos em Elixir.
```
Comment (vbrazo) [inline pt/lessons/libraries/mox.md:6]:
```suggestion
Mox é uma biblioteca feita para projetar mocks simultâneos em Elixir.
```

File: pt/lessons/libraries/mox.md [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+
+{% include toc.html %}
+
+## Escrever Código Testável
+
+Os testes e os mocks que os facilitam não são, normalmente, o destaque de qualquer língua, e, por isso, não é surpreendente que exista menos literatura sobre eles.
+No entanto, pode-se _absolutamente_ usar mocks em Elixir!
+A metodologia exacta pode ser um pouco diferente da que está familiarizado noutras linguas, mas o objetivo final é o mesmo: os mocks podem simular o output de funções internas e, então, permitem-lhe assumir todas as possíveis execuções do seu código.
```
Comment (vbrazo) [inline pt/lessons/libraries/mox.md:14]:
```suggestion
A metodologia exacta pode ser um pouco diferente da que está familiarizado noutras línguas, mas o objetivo final é o mesmo: os mocks podem simular o output de funções internas e, então, permitem-lhe assumir todas as possíveis execuções do seu código.
```

File: pt/lessons/libraries/mox.md [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+config :my_app, :http_client, HTTPoison
+```
+
+Esta lógica e a sua dependência no config da aplicação forma a base de tudo que se segue.
+
+Se você é propenso a pensar em demasia, sim, poderia omitir a função `http_client/0` e chamar diretamente `Application.get_env/2`, e, sim, poderia também fornecer um terceiro argumento padrão a `Application.get_env/3` e obter o mesmo resultado.
+
+Aproveitando-nos do config da aplicação, é-nos permitido ter implementações específicas do módulo para cada ambiente; você poderia fazer referência a um módulo sandbox para o ambiente `dev` enquanto que o ambiente `test` poderia usar um módulo da memória.
+
+No entanto, ter um único módulo fico por ambiente pode não ser flexível o suficiente: dependendo de com a sua função é usada, você poderá necessitar de devolver diferentes respostas para conseguir testar todos os caminhos de execução possíveis.
```
Comment (vbrazo) [inline pt/lessons/libraries/mox.md:75]:
```suggestion
No entanto, ter um único módulo fico por ambiente pode não ser flexível o suficiente: dependendo de como a sua função é usada, você poderá necessitar de devolver diferentes respostas para conseguir testar todos os caminhos de execução possíveis.
```

File: pt/lessons/libraries/mox.md [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+Mox.defmock(HTTPoison.BaseMock, for: HTTPoison.Base)
+# ... etc...
+
+# 3. Dar override das configurações do config (similar a adicionar os mesmos a `config/test.exs`)
+Application.put_env(:my_app, :http_client, HTTPoison.BaseMock)
+# ... etc...
+```
+
+Algumas coisas importantes a fazerem-se notar sobre `Mox.defmock`: o nome do lado esquerdo é arbitrário.
+Os nomes dos módulos em Elixir são apenas atoms -- você não precisa de criar o módulo em lado algum, tudo que precisa de fazer é "reservar" o nome para um módulo mock.
```
Comment (vbrazo) [inline pt/lessons/libraries/mox.md:180]:
```suggestion
Os nomes dos módulos em Elixir são apenas atoms -- você não precisa criar o módulo em lado algum, tudo que precisa de fazer é "reservar" o nome para um módulo mock.
```

File: pt/lessons/libraries/mox.md [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+Nos bastidores, o Mox irá criar o módulo com esse nome em tempo real dentro do BEAM.
+
+A segunda coisa complicada é que o módulo referido por `for:` _deve_ ser um comportamento: _deve_ definir callbacks.
+Mox usa introspecção neste módulo e você poderá apenas definir funções mock quando uma `@callback` tiver sido definida.
+É assim que o Mox aplica o contrato.
+Às vezes é pode ser difícil encontrar o módulo de comportamento: `HTTPoison`, por exemplo, depende do `HTTPoison.Base`, mas você não tem forma de saber isso a não ser que olhe para o source code.
```
Comment (vbrazo) [inline pt/lessons/libraries/mox.md:186]:
```suggestion
Às vezes pode ser difícil encontrar o módulo de comportamento: `HTTPoison`, por exemplo, depende do `HTTPoison.Base`, mas você não tem forma de saber isso a não ser que olhe para o source code.
```

File: pt/lessons/libraries/mox.md [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+    expect(HTTPoison.BaseMock, :get, fn _ -> {:error, "Sorry!"} end)
+    assert {:error, _} = MyModule.get_username("does-not-exist")
+  end
+end
+```
+
+Para cada teste, nós fazemos referência ao _mesmo_ módulo mock (`HTTPoison.BaseMock` neste exemplo), e usamos a função `expect` para definir o valor que é retornado para cada função chamada.
+
+Usar o `Mox` é perfeitamente seguro para uma execução assíncrona, e requer que cada mock siga um contrato.
+Atendendo que estes mocks são "virtuais", não há necessidade de definir módulos reais que poderia atrapalhar a nossa aplicação.
```
Comment (vbrazo) [inline pt/lessons/libraries/mox.md:220]:
```suggestion
Atendendo que estes mocks são "virtuais", não há necessidade de definir módulos reais que poderiam atrapalhar a nossa aplicação.
```

**Comment by ItzaMi** [comment]
Thank you for the changes, @vbrazo 🤘

File: pt/lessons/libraries/mox.md [hunk 7]
```diff
@@ -0,0 +1,222 @@
+---
+version: 1.0.0
+title: Mox
+---
+
+Mox é uma biblioteca feita para projetar mocks simultâneos em Elixir.
```
Comment (thiamsantos) [inline pt/lessons/libraries/mox.md:6]:
```suggestion
Mox é uma biblioteca feita para criar mocks concorrentes em Elixir.
```
WDYT?

File: pt/lessons/libraries/mox.md [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+
+{% include toc.html %}
+
+## Escrever Código Testável
+
+Os testes e os mocks que os facilitam não são, normalmente, o destaque de qualquer língua, e, por isso, não é surpreendente que exista menos literatura sobre eles.
```
Comment (thiamsantos) [inline pt/lessons/libraries/mox.md:12]:
```suggestion
Os testes e os mocks que os facilitam não são, normalmente, o destaque de qualquer linguagem, e, por isso, não é surpreendente que exista menos literatura sobre eles.
```

File: pt/lessons/libraries/mox.md [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+No entanto, pode-se _absolutamente_ usar mocks em Elixir!
```
Comment (thiamsantos) [inline pt/lessons/libraries/mox.md:13]:
```suggestion
No entanto, você pode _absolutamente_ usar mocks em Elixir!
```

File: pt/lessons/libraries/mox.md [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+A metodologia exacta pode ser um pouco diferente da que está familiarizado noutras línguas, mas o objetivo final é o mesmo: os mocks podem simular o output de funções internas e, então, permitem-lhe assumir todas as possíveis execuções do seu código.
```
Comment (thiamsantos) [inline pt/lessons/libraries/mox.md:14]:
```suggestion
A metodologia exata pode ser um pouco diferente da que você está familiarizado em outras linguagens, mas o objetivo final é o mesmo: os mocks podem simular o output de funções internas e, então, permitem-lhe verificar todas os possíveis caminhos de execução do seu código.
```

File: pt/lessons/libraries/mox.md [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+
+Antes de mostrarmos use cases complexos, vamos falar de algumas técnicas que lhe poderão ajudar a tornar o seu código mais testável.
```
Comment (thiamsantos) [inline pt/lessons/libraries/mox.md:16]:
```suggestion
Antes de vermos casos de uso mais complexos, vamos falar de algumas técnicas que podem nos ajudar a tornar o nosso código mais testável.
```

File: pt/lessons/libraries/mox.md [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+Uma tática simples é passar um módulo para uma funções em vez de fazer hard-code do módulo dentro da função.
```
Comment (thiamsantos) [inline pt/lessons/libraries/mox.md:17]:
```suggestion
Uma tática simples é passar um módulo para uma função em vez de ter o módulo hard-coded dentro da função.
```
I'm not sure in the approach here. Would be nice if we had a translation for "hard-coding".

File: pt/lessons/libraries/mox.md [hunk 13]
```diff
@@ -0,0 +1,222 @@
...
+
+Por exemplo, se fizessemos hard-code de um cliente HTTP dentro de uma função:
```
Comment (thiamsantos) [inline pt/lessons/libraries/mox.md:19]:
```suggestion
Por exemplo, se escrevermos um cliente HTTP diretamente dentro de uma função:
```

WDYT?

File: pt/lessons/libraries/mox.md (hunk 12 above)
Comment (ItzaMi) [reply pt/lessons/libraries/mox.md:17]: I tried to google for portuguese articles with `hard-coded` in it to see possible translations but really couldn't find anything suitable

Comment (ItzaMi) [reply pt/lessons/libraries/mox.md:17]: Maybe `Uma tática simples é passar um módulo para uma função em vez de ter o módulo diretamente dentro da função.` but I don't think it properly passes the intended message

File: pt/lessons/libraries/mox.md [hunk 14]
```diff
@@ -0,0 +1,222 @@
...
+
+Mox é uma biblioteca feita para criar mocks concorrentes em Elixir.
+
+{% include toc.html %}
+
+## Escrever Código Testável
+
+Os testes e os mocks que os facilitam não são, normalmente, o destaque de qualquer linguagem, e, por isso, não é surpreendente que exista menos literatura sobre eles.
+No entanto, você pode _absolutamente_ usar mocks em Elixir!
+A metodologia exata pode ser um pouco diferente da que você está familiarizado em outras linguagens, mas o objetivo final é o mesmo: os mocks podem simular o output de funções internas e, então, permitem-lhe verificar todas os possíveis caminhos de execução do seu código.
```
Comment (vbrazo) [inline pt/lessons/libraries/mox.md:14]:
```suggestion
A metodologia exata pode ser um pouco diferente da que você está familiarizado em outras linguagens, mas o objetivo final é o mesmo: os mocks podem simular o output de funções internas e, então, permitem-lhe verificar todos os possíveis caminhos de execução do seu código.
```

File: pt/lessons/libraries/mox.md (hunk 12 above)
Comment (vbrazo) [reply pt/lessons/libraries/mox.md:17]: I remember a long time ago, in Brazil, a friend of mine used to say `engessado` that comes from "teto de gesso" lol, but I don't think it's applicable for this context. I'd probably consider leaving `hard-coded` for now and it should be ok.

Comment (ItzaMi) [reply pt/lessons/libraries/mox.md:17]: Sounds good to me, I completely agree with that!

File: pt/lessons/libraries/mox.md [hunk 15]
```diff
@@ -0,0 +1,222 @@
...
+
+Ou poderíamos usar a função [apply/3](https://hexdocs.pm/elixir/Kernel.html#apply/3) que realiza o mesmo:
+
+```elixir
+def get_username(username, http_client) do
+  apply(http_client, :get, ["https://elixirschool.com/users/#{username}"])
+end
+```
+
+Passar o módulo como argumento ajuda a separar as eventuais preocupações que tenhamos e, se não nos assustarmos demasiado com a verborragia de programação orientada a objetos na definição, poderemos reconhecer esta inversão de controlo como uma espécie de [Injeção de Dependência](https://en.wikipedia.org/wiki/Dependency_injection).
```
Comment (thiamsantos) [inline pt/lessons/libraries/mox.md:43]:
```suggestion
Passar o módulo como argumento ajuda a separar as responsabilidades e, se não nos assustarmos demais com a verbosidade de programação orientada a objetos na definição, poderemos reconhecer esta inversão de controle como uma espécie de [Injeção de Dependência](https://en.wikipedia.org/wiki/Dependency_injection).
```

File: pt/lessons/libraries/mox.md [hunk 16]
```diff
@@ -0,0 +1,222 @@
...
+Para testar o método `get_username/2`, você só necessitaria de passar um módulo custa função `get` retorne o valor necessário para as suas asserções.
```
Comment (thiamsantos) [inline pt/lessons/libraries/mox.md:44]:
```suggestion
Para testar o método `get_username/2`, você só precisaria passar um módulo com uma função `get` que retorne o valor necessário para as suas verificações.
```

File: pt/lessons/libraries/mox.md [hunk 17]
```diff
@@ -0,0 +1,222 @@
...
+
+Esta lógica é muito simples, e, por isso, é apenas útil quando a função é facilmente acessível (e não, por exemplo, quando está subterrada algures dentro de uma função privada).
```
Comment (thiamsantos) [inline pt/lessons/libraries/mox.md:46]:
```suggestion
Esta lógica é muito simples, e, por isso, é apenas útil quando a função é facilmente acessível (e não, por exemplo, quando está enterrada em algum lugar bem fundo no uma função privada).
```

File: pt/lessons/libraries/mox.md [hunk 18]
```diff
@@ -0,0 +1,222 @@
...
+
+Uma tática mais flexível apoia-se na configuração da aplicação.
```
Comment (thiamsantos) [inline pt/lessons/libraries/mox.md:48]:
```suggestion
Uma tática mais flexível depende de configuração da aplicação.
```

File: pt/lessons/libraries/mox.md [hunk 19]
```diff
@@ -0,0 +1,222 @@
...
+Talvez não se tenha ainda apercebido, mas uma aplicação Elixir mantém o estado na sua configuração.
+Em vez de fazer hard-code de um módulo ou passá-lo como um argumento, pode ler o mesmo a partir da configuração da aplicação.
+Perhaps you didn't even realize it, but an Elixir application maintains state in its configuration.
```
Comment (thiamsantos) [inline pt/lessons/libraries/mox.md:51]: :scissors:

File: pt/lessons/libraries/mox.md [hunk 20]
```diff
@@ -0,0 +1,222 @@
...
+def get_username(username) do
+  http_client().get("https://elixirschool.com/users/#{username}")
+end
+
+defp http_client do
+  Application.get_env(:my_app, :http_client)
+end
+```
+
+Então, no seu ficheiro config:
```
Comment (thiamsantos) [inline pt/lessons/libraries/mox.md:63]:
```suggestion
Então, no seu arquivo de configuração:
```

File: pt/lessons/libraries/mox.md [hunk 21]
```diff
@@ -0,0 +1,222 @@
...
+
+```elixir
+config :my_app, :http_client, HTTPoison
+```
+
+Esta lógica e a sua dependência no config da aplicação forma a base de tudo que se segue.
```
Comment (thiamsantos) [inline pt/lessons/libraries/mox.md:69]:
```suggestion
Esta lógica e a sua dependência na configuração da aplicação forma a base de tudo que se segue.
```

File: pt/lessons/libraries/mox.md [hunk 22]
```diff
@@ -0,0 +1,222 @@
...
+
+Se você é propenso a pensar em demasia, sim, poderia omitir a função `http_client/0` e chamar diretamente `Application.get_env/2`, e, sim, poderia também fornecer um terceiro argumento padrão a `Application.get_env/3` e obter o mesmo resultado.
```
Comment (thiamsantos) [inline pt/lessons/libraries/mox.md:71]:
```suggestion
Se você é propenso a pensar demais, sim, poderia omitir a função `http_client/0` e chamar diretamente `Application.get_env/2`, e, sim, poderia também fornecer um terceiro argumento padrão a `Application.get_env/3` e obter o mesmo resultado.
```

File: pt/lessons/libraries/mox.md [hunk 23]
```diff
@@ -0,0 +1,222 @@
...
+
+Aproveitando-nos do config da aplicação, é-nos permitido ter implementações específicas do módulo para cada ambiente; você poderia fazer referência a um módulo sandbox para o ambiente `dev` enquanto que o ambiente `test` poderia usar um módulo da memória.
```
Comment (thiamsantos) [inline pt/lessons/libraries/mox.md:73]:
```suggestion
Nos aproveitar do configuração da aplicação nos permite ter implementações específicas do módulo para cada ambiente; você poderia fazer referência a um módulo sandbox para o ambiente `dev` enquanto que o ambiente `test` poderia usar um módulo da memória.
```

File: pt/lessons/libraries/mox.md [hunk 24]
```diff
@@ -0,0 +1,222 @@
...
+
+No entanto, ter um único módulo fico por ambiente pode não ser flexível o suficiente: dependendo de com a sua função é usada, você poderá necessitar de devolver diferentes respostas para conseguir testar todos os caminhos de execução possíveis.
```
Comment (thiamsantos) [inline pt/lessons/libraries/mox.md:75]:
```suggestion
Contudo, ter um único módulo fixo por ambiente pode não ser flexível o suficiente: dependendo de como a sua função é usada, você pode precisar devolver diferentes respostas para conseguir testar todos os caminhos de execução possíveis.
```

File: pt/lessons/libraries/mox.md [hunk 25]
```diff
@@ -0,0 +1,222 @@
...
+O que a maior parte das pessoas não sabe é que você pode _mudar_ a configuração da aplicação em templo de execução!
+Vamos dar uma olhada ao [Application.put_env/4](https://hexdocs.pm/elixir/Application.html#put_env/4).
```
Comment (thiamsantos) [inline pt/lessons/libraries/mox.md:77]:
```suggestion
Vamos dar uma olhada em [Application.put_env/4](https://hexdocs.pm/elixir/Application.html#put_env/4).
```

File: pt/lessons/libraries/mox.md [hunk 26]
```diff
@@ -0,0 +1,222 @@
...
+
+Imagine que a sua aplicação necessita de agir de forma diferente dependendo de se o pedido HTTP foi, ou não, feito com sucesso.
```
Comment (thiamsantos) [inline pt/lessons/libraries/mox.md:79]:
```suggestion
Imagine que a sua aplicação precisa agir de forma diferente dependendo de se a requisição HTTP foi, ou não, feita com sucesso.
```

File: pt/lessons/libraries/mox.md [hunk 27]
```diff
@@ -0,0 +1,222 @@
...
+Poderíamos criar múltiplos módulos, cada um com uma função `get/1`.
+Um módulo poderia devolver um tuple `:ok`, e outro poderia devolver um tuple `:error`.
```
Comment (thiamsantos) [inline pt/lessons/libraries/mox.md:81]:
```suggestion
Um módulo poderia devolver uma tupla `:ok`, e outro poderia devolver uma tupla `:error`.
```

File: pt/lessons/libraries/mox.md [hunk 28]
```diff
@@ -0,0 +1,222 @@
...
+  end
+
+  test ":error on 404" do
+    Application.put_env(:my_app, :http_client, HTTP404Mock)
+    assert {:error, _} = MyModule.get_username("does-not-exist")
+  end
+end
+```
+
+É assumido que você tenha criado os módulos necessários algures (`HTTP200Mock` e `HTTP404Mock`).
```
Comment (thiamsantos) [inline pt/lessons/libraries/mox.md:111]:
```suggestion
É assumido que você tenha criado os módulos necessários em algum lugar (`HTTP200Mock` e `HTTP404Mock`).
```

File: pt/lessons/libraries/mox.md [hunk 29]
```diff
@@ -0,0 +1,222 @@
...
+Nós adicionamos um callback [`on_exit`](https://hexdocs.pm/ex_unit/master/ExUnit.Callbacks.html#on_exit/2) ao [`setup`](https://hexdocs.pm/ex_unit/master/ExUnit.Callbacks.html#setup/1) para assegurar que o `:http_client` é devolvido ao seu estado anterior depois de cada teste.
+
+No entanto, um padrão como o descrito acima normalmente _NÃO_ é algo que você deva seguir ou fazer!
+As razãos para isto poderão não ser imediatamente óbvias.
+
+Primeiramente, não há nada que garante que os módulos que definimos no nosso `:http_client` podem fazer o que necessitam de fazer: não há aqui a imposição de um contrato que requer que os módulos tenham uma função `get/1`.
+
+Em segundo lugar, testes como o descrito acima não podem ser corridos com segurança de forma assíncrona.
```
Comment (thiamsantos) [inline pt/lessons/libraries/mox.md:119]:
```suggestion
Em segundo lugar, testes como o descrito acima não podem rodar com segurança de forma assíncrona.
```

File: pt/lessons/libraries/mox.md [hunk 30]
```diff
@@ -0,0 +1,222 @@
...
+Devido ao estado da aplicação ser partilhado por _toda_ a aplicação, é completamente possível que quando você dá override do `:http_client` num teste, que um outro teste (a correr em simultâneo) espere um resultado diferente.
```
Comment (thiamsantos) [inline pt/lessons/libraries/mox.md:120]:
```suggestion
Devido ao estado da aplicação ser partilhado por _toda_ a aplicação, é completamente possível que quando você dá override do `:http_client` num teste, que um outro teste (rodando simultaneamente) espere um resultado diferente.
```

File: pt/lessons/libraries/mox.md [hunk 31]
```diff
@@ -0,0 +1,222 @@
...
+Você pode ter encontrado problemas como este quando o teste é executado _usualmente_ passa, mas às vezes falha inexplicavelmente. Cuidado!
+
+Em terceiro lugar, esta abordagem pode ficar confusa porque você acaba com um conjunto de módulos mock algures no meio da sua aplicação. Que nojo.
```
Comment (thiamsantos) [inline pt/lessons/libraries/mox.md:123]:
```suggestion
Em terceiro lugar, esta abordagem pode ficar confusa porque você acaba com um conjunto de módulos mock em algum lugar no meio da sua aplicação. Que nojo.
```

File: pt/lessons/libraries/mox.md [hunk 32]
```diff
@@ -0,0 +1,222 @@
...
+
+Fizemos uma demonstração da estrutura acima porque descreve a abordagem de uma forma bastante direta que nos ajuda a compreender um pouco melhor sobre como a solução _real_ funciona.
+
+## Mox : A Solução para todos os Problemas
+
+A package ideal para trabalhar com mocks em Elixir é a [Mox](https://hexdocs.pm/mox/Mox.html), da autoria do prório José Valim, e resolve todos os problemas delineados acima.
```
Comment (thiamsantos) [inline pt/lessons/libraries/mox.md:129]:
```suggestion
A bibiloteca ideal para trabalhar com mocks em Elixir é a [Mox](https://hexdocs.pm/mox/Mox.html), de autoria do próprio José Valim, e resolve todos os problemas delineados acima.
```

File: pt/lessons/libraries/mox.md [hunk 33]
```diff
@@ -0,0 +1,222 @@
...
+
+Lembre-se: como requisito, o seu código deve olhar para o config da sua aplicação para obter o seu módulo configurado:
```
Comment (thiamsantos) [inline pt/lessons/libraries/mox.md:131]:
```suggestion
Lembre-se: como requisito, o seu código deve olhar para a configuração da sua aplicação para obter o seu módulo configurado:
```

File: pt/lessons/libraries/mox.md [hunk 34]
```diff
@@ -0,0 +1,222 @@
...
+end
+```
+
+Instale o mesmo com `mix deps.get`.
+
+Depois, modifique o seu `test_helper.exs` para que este faça 3 coisas:
+
+1. deve começar a app Mox
+2. deve definir um ou mais mocks
+3. deve definir a configuração do aplicativo com o mock
```
Comment (thiamsantos) [inline pt/lessons/libraries/mox.md:161]:
```suggestion
3. deve definir a configuração da aplicação com o mock
```

File: pt/lessons/libraries/mox.md [hunk 35]
```diff
@@ -0,0 +1,222 @@
...
+# 2. definir mocks dinâmicos
+Mox.defmock(HTTPoison.BaseMock, for: HTTPoison.Base)
+# ... etc...
+
+# 3. Dar override das configurações do config (similar a adicionar os mesmos a `config/test.exs`)
+Application.put_env(:my_app, :http_client, HTTPoison.BaseMock)
+# ... etc...
+```
+
+Algumas coisas importantes a fazerem-se notar sobre `Mox.defmock`: o nome do lado esquerdo é arbitrário.
```
Comment (thiamsantos) [inline pt/lessons/libraries/mox.md:179]:
```suggestion
Algumas coisas importantes a se notar sobre o `Mox.defmock`: o nome do lado esquerdo é arbitrário.
```

File: pt/lessons/libraries/mox.md [hunk 36]
```diff
@@ -0,0 +1,222 @@
...
+Os nomes dos módulos em Elixir são apenas atoms -- você não precisa criar o módulo em lado algum, tudo que precisa de fazer é "reservar" o nome para um módulo mock.
```
Comment (thiamsantos) [inline pt/lessons/libraries/mox.md:180]:
```suggestion
Os nomes dos módulos em Elixir são apenas atoms -- você não precisa criar o módulo em nenhum, tudo que você está fazendo é "reservando" um nome para um módulo mock.
```

File: pt/lessons/libraries/mox.md [hunk 37]
```diff
@@ -0,0 +1,222 @@
...
+Nos bastidores, o Mox irá criar o módulo com esse nome em tempo real dentro do BEAM.
+
+A segunda coisa complicada é que o módulo referido por `for:` _deve_ ser um comportamento: _deve_ definir callbacks.
+Mox usa introspecção neste módulo e você poderá apenas definir funções mock quando uma `@callback` tiver sido definida.
+É assim que o Mox aplica o contrato.
+Às vezes pode ser difícil encontrar o módulo de comportamento: `HTTPoison`, por exemplo, depende do `HTTPoison.Base`, mas você não tem forma de saber isso a não ser que olhe para o source code.
```
Comment (thiamsantos) [inline pt/lessons/libraries/mox.md:186]:
```suggestion
Às vezes pode ser difícil encontrar o módulo de comportamento: `HTTPoison`, por exemplo, depende do `HTTPoison.Base`, mas você não tem forma de saber isso a não ser que olhe para o código fonte.
```

File: pt/lessons/libraries/mox.md [hunk 38]
```diff
@@ -0,0 +1,222 @@
...
+Se estiver a tentar criar um mock para uma 3rd-party package, poderá já ter descoberto que esse comportamento não existe!
```
Comment (thiamsantos) [inline pt/lessons/libraries/mox.md:187]:
```suggestion
Se estiver a tentar criar um mock para uma biblioteca de terceira, poderá já ter descoberto que esse comportamento não existe!
```

File: pt/lessons/libraries/mox.md [hunk 39]
```diff
@@ -0,0 +1,222 @@
...
+Nesses casos você poderá ter de definir o seu próprio comportamento e callbacks para satisfazer a necessidade de um contrato.
+
+Esta situação traz consigo um ponto importante: você poderá querer usar uma camada de abstração (ou seja, [indirection](https://en.wikipedia.org/wiki/Indirection)) para que a sua aplicação não dependa de um pacto de third-party _diretamente_, mas, em vez disso, você usaria o seu próprio módulo que, por usa vez, usaria essa package.
```
Comment (thiamsantos) [inline pt/lessons/libraries/mox.md:190]:
```suggestion
Esta situação traz consigo um ponto importante: você poderá querer usar uma camada de abstração (ou seja, [indirection](https://en.wikipedia.org/wiki/Indirection)) para que a sua aplicação não dependa de uma biblioteca de terceiro _diretamente_, mas, em vez disso, você usaria o seu próprio módulo que, por usa vez, usaria essa biblioteca.
```

File: pt/lessons/libraries/mox.md [hunk 40]
```diff
@@ -0,0 +1,222 @@
...
+É importante, numa aplicação bem desenhada e concebida, definir os 'limites' adequados, mas a mecânica dos mocks não se altera, por isso, não deixe que isso o atrapalhe.
```
Comment (thiamsantos) [inline pt/lessons/libraries/mox.md:191]:
```suggestion
É importante, numa aplicação bem desenhada e concebida, definir os "limites" adequados, mas a mecânica dos mocks não se altera, por isso, não deixe que isso o atrapalhe.
```

**Comment by ItzaMi** [comment]
@thiamsantos, I appreciate the fact that you're taking so much time to review it!

File: pt/lessons/libraries/mox.md [hunk 41]
```diff
@@ -0,0 +1,221 @@
...
+```elixir
+def get_username(username, http_client) do
+  apply(http_client, :get, ["https://elixirschool.com/users/#{username}"])
+end
+```
+
+Passar o módulo como argumento ajuda a separar as responsabilidades e, se não nos assustarmos demais com a verbosidade de programação orientada a objetos na definição, poderemos reconhecer esta inversão de controle como uma espécie de [Injeção de Dependência](https://en.wikipedia.org/wiki/Dependency_injection).
+Para testar o método `get_username/2`, você só precisaria passar um módulo com uma função `get` que retorne o valor necessário para as suas verificações.
+
+Esta lógica é muito simples, e, por isso, é apenas útil quando a função é facilmente acessível (e não, por exemplo, quando está enterrada em algum lugar bem fundo no uma função privada).
```
Comment (thiamsantos) [inline pt/lessons/libraries/mox.md:46]:
```suggestion
Esta lógica é muito simples, e, por isso, é apenas útil quando a função é facilmente acessível (e não, por exemplo, quando está enterrada em algum lugar bem fundo de uma função privada).
```

**Review by thiamsantos — APPROVED** [review]

**Comment by thiamsantos** [comment]
Thanks @ItzaMi :heart: :tada:

**Pull request merged**